File: app/src/main/java/com/github/ilikeyourhat/sudokusolver/solving/deduction/combinations/CollectionCombinator.java
package com.github.ilikeyourhat.sudokusolver.solving.deduction.combinations;

import java.util.*;

public class CollectionCombinator {

    private final int combinationSize;

    public CollectionCombinator(int combinationSize) {
        this.combinationSize = combinationSize;
    }

    @SuppressWarnings("unchecked")
    public <T> void iterate(Collection<T> collection, CombinationConsumer<T> consumer) {
        if (collection.size() < combinationSize) return;
        CombinationIterator iterator = new CombinationIterator(combinationSize, collection.size());
        T[] elements = (T[]) collection.toArray();
        while (iterator.hasNext()) {
            List<T> values = new ArrayList<>();
            for (int index : iterator.next()) {
                T value = elements[index];
                values.add(value);
            }
            consumer.onNextCombination(values);
        }
    }

    @FunctionalInterface
    public interface CombinationConsumer<T> {
        void onNextCombination(List<T> values);
    }

}

File: app/src/main/kotlin/com/github/ilikeyourhat/kudoku/solving/deduction/combinations/CollectionCombinator.kt
package com.github.ilikeyourhat.kudoku.solving.deduction.combinations

class CollectionCombinator(private val combinationSize: Int) {

    fun <T> iterate(collection: Collection<T>, consumer: CombinationConsumer<T>) {
        if (collection.size < combinationSize) return
        val iterator = CombinationIterator(combinationSize, collection.size)
        while (iterator.hasNext()) {
            val values: MutableList<T> = ArrayList()
            for (index in iterator.next()) {
                val value = collection.toList()[index]
                values.add(value)
            }
            consumer.onNextCombination(values)
        }
    }

    fun interface CombinationConsumer<T> {
        fun onNextCombination(values: List<T>)
    }
}
